File: common/scala/src/main/scala/org/aiddl/common/scala/planning/state_variable/GoalTest.scala
package org.aiddl.common.scala.planning.state_variable

import scala.collection.mutable.Set
import scala.collection.mutable.Map
import scala.collection.mutable.HashMap
import scala.collection.mutable.HashSet

import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.function.Initializable
import org.aiddl.core.scala.function.Configurable

import org.aiddl.core.scala.container.Container

import org.aiddl.core.scala.representation._

import org.aiddl.common.scala.planning.PlanningTerm._

import Term.given_Conversion_Term_KeyVal
import scala.language.implicitConversions

class GoalTest extends Function with Initializable {
  var goal: SetTerm = SetTerm()

  def init( g: Term ): Unit = { this.goal = g.asSet }

  def apply( s: Term ): Term = Bool(goal.forall( g => s.get(g.key) match {
    case Some(v) => v == g.value
    case None => false } ))
}

File: common/java/src/main/java/org/aiddl/common/java/planning/state_variable/GoalTest.java
package org.aiddl.common.java.planning.state_variable;

import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.interfaces.InitializableFunction;
import org.aiddl.core.java.interfaces.InterfaceImplementation;
import org.aiddl.core.java.representation.SetTerm;
import org.aiddl.core.java.representation.SymbolicTerm;
import org.aiddl.core.java.representation.Term;

public class GoalTest implements Function, InitializableFunction, InterfaceImplementation {
	SetTerm goal;
	
	private final static SymbolicTerm InterfaceUri = Term.sym("org.aiddl.common.planning.state-variable.goal-test");
	
	@Override
	public SymbolicTerm getInterfaceUri() {
		return InterfaceUri;
	}
		
	@Override
	public void initialize(Term args) {
		this.goal = args.asSet();
	}	
	
	@Override
	public Term apply(Term args) {
		for ( Term g : goal ) {
			Term s_a = args.asSet().get(g.getKey());
			if ( s_a == null || !s_a.equals(g.getValue()) ) {
				return Term.bool(false);
			}
		}
		return Term.bool(true);
	}
}
